File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueAproximacaoLinear.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Aproximação linear (criptoanálise de Matsui): procura por correlação entre
 * bits de entrada (IV) e bits de saída do keystream. Uma cifra ideal não tem
 * aproximação linear com viés relevante; bias alto é uma pista explorável.
 */
public class AtaqueAproximacaoLinear implements AtaqueInterface {
    private final int amostras;
    private final int tamanhoBloco;
    private final double limiteBias;

    public AtaqueAproximacaoLinear() {
        this(200, 32, 0.3);
    }

    public AtaqueAproximacaoLinear(int amostras, int tamanhoBloco, double limiteBias) {
        this.amostras = amostras;
        this.tamanhoBloco = tamanhoBloco;
        this.limiteBias = limiteBias;
    }

    @Override
    public String nome() {
        return "Aproximação linear (viés de Walsh)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        int tamIv = alvo.tamanhoIv();
        int bitsEntrada = tamIv * 8;
        int bitsSaida = this.tamanhoBloco * 8;
        int lenChave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8).length;

        int[][] entradas = new int[this.amostras][bitsEntrada];
        int[][] saidas = new int[this.amostras][bitsSaida];

        for (int s = 0; s < this.amostras; s++) {
            byte[] chave = Util.randomBytes(lenChave);
            byte[] iv = Util.randomBytes(tamIv);
            byte[] ks = alvo.gerarKeystreamBruto(chave, iv, "enc", this.tamanhoBloco);

            for (int j = 0; j < bitsEntrada; j++) {
                entradas[s][j] = (iv[j >> 3] >> (7 - (j & 7))) & 1;
            }
            for (int j = 0; j < bitsSaida; j++) {
                saidas[s][j] = (ks[j >> 3] >> (7 - (j & 7))) & 1;
            }
        }

        double maiorBias = 0;
        int[] par = {-1, -1};
        for (int i = 0; i < bitsEntrada; i++) {
            for (int j = 0; j < bitsSaida; j++) {
                int iguais = 0;
                for (int s = 0; s < this.amostras; s++) {
                    if (entradas[s][i] == saidas[s][j]) {
                        iguais++;
                    }
                }
                double bias = Math.abs((double) iguais / this.amostras - 0.5);
                if (bias > maiorBias) {
                    maiorBias = bias;
                    par = new int[] {i, j};
                }
            }
        }

        boolean vulneravel = maiorBias > this.limiteBias;

        Map<String, Object> dados = new HashMap<>();
        dados.put("maior_bias", maiorBias);
        dados.put("par", new int[] {par[0], par[1]});

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.ALTA : Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "maior viés |p-0.5|=%.4f na aproximação IV[bit %d] -> saída[bit %d] (limite %.2f); %d amostras",
                        maiorBias,
                        par[0],
                        par[1],
                        this.limiteBias,
                        this.amostras),
                dados);
    }
}

File: packages/go/internal/seguranca/ataques/ataque_aproximacao_linear.go
package ataques

import (
	"fmt"
	"math"

	"criptografia/internal/seguranca"
)

// OpcoesAproximacaoLinear configura o AtaqueAproximacaoLinear.
type OpcoesAproximacaoLinear struct {
	Amostras     int
	TamanhoBloco int
	LimiteBias   float64
}

// AtaqueAproximacaoLinear (criptoanálise de Matsui) procura correlação entre
// bits de entrada (IV) e bits de saída do keystream.
type AtaqueAproximacaoLinear struct {
	amostras     int
	tamanhoBloco int
	limiteBias   float64
}

// NovoAtaqueAproximacaoLinear cria o ataque. Sem argumento, usa amostras = 200,
// tamanhoBloco = 32 e limiteBias = 0.3.
func NovoAtaqueAproximacaoLinear(opcoes ...OpcoesAproximacaoLinear) *AtaqueAproximacaoLinear {
	o := OpcoesAproximacaoLinear{Amostras: 200, TamanhoBloco: 32, LimiteBias: 0.3}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueAproximacaoLinear{amostras: o.Amostras, tamanhoBloco: o.TamanhoBloco, limiteBias: o.LimiteBias}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueAproximacaoLinear) Nome() string {
	return "Aproximação linear (viés de Walsh)"
}

// Executar roda o ataque.
func (a *AtaqueAproximacaoLinear) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	tamIv := alvo.TamanhoIv()
	bitsEntrada := tamIv * 8
	bitsSaida := a.tamanhoBloco * 8

	entradas := make([][]byte, a.amostras)
	saidas := make([][]byte, a.amostras)

	for s := 0; s < a.amostras; s++ {
		chave := seguranca.RandomBytes(len(alvo.ChaveDeTeste()))
		iv := seguranca.RandomBytes(tamIv)
		ks := alvo.GerarKeystreamBruto(chave, iv, "enc", a.tamanhoBloco)

		inBits := make([]byte, bitsEntrada)
		for j := 0; j < bitsEntrada; j++ {
			inBits[j] = (iv[j>>3] >> (7 - (j & 7))) & 1
		}
		outBits := make([]byte, bitsSaida)
		for j := 0; j < bitsSaida; j++ {
			outBits[j] = (ks[j>>3] >> (7 - (j & 7))) & 1
		}
		entradas[s] = inBits
		saidas[s] = outBits
	}

	maiorBias := 0.0
	par := [2]int{-1, -1}
	for i := 0; i < bitsEntrada; i++ {
		for j := 0; j < bitsSaida; j++ {
			iguais := 0
			for s := 0; s < a.amostras; s++ {
				if entradas[s][i] == saidas[s][j] {
					iguais++
				}
			}
			bias := math.Abs(float64(iguais)/float64(a.amostras) - 0.5)
			if bias > maiorBias {
				maiorBias = bias
				par = [2]int{i, j}
			}
		}
	}

	vulneravel := maiorBias > a.limiteBias
	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevAlta
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"maior viés |p-0.5|=%.4f na aproximação IV[bit %d] -> saída[bit %d] (limite %v); %d amostras",
			maiorBias, par[0], par[1], a.limiteBias, a.amostras,
		),
		Dados: map[string]any{"maior_bias": maiorBias, "par": par},
	}, nil
}
